src/middlewared/middlewared/plugins/mail.py

```diff
@@ -1,41 +1,324 @@
         Str('channel'),
-        Int('timeout'),
+        Int('timeout', default=300),
+        Bool('attachments', default=False),
         Bool('queue', default=True),
-    ))
-    def send(self, message):
+        Dict('extra_headers', additional_attrs=True),
+    ), Dict('mailconfig', additional_attrs=True))
+    @job(pipe=True)
+    def send(self, job, message, config=None):
         """
         Sends mail using configured mail settings.
+
+        If `attachments` is true, a list compromised of the following dict is required
+        via HTTP upload:
+          - headers(list)
+            - name(str)
+            - value(str)
+            - params(dict)
+          - content (str)
+
+        [
+         {
+          "headers": [
+           {
+            "name": "Content-Transfer-Encoding",
+            "value": "base64"
+           },
+           {
+            "name": "Content-Type",
+            "value": "application/octet-stream",
+            "params": {
+             "name": "test.txt"
+            }
+           }
+          ],
+          "content": "dGVzdAo="
+         }
+        ]
         """
-        # TODO: For now this is just a wrapper for freenasUI send_mail,
-        #       when the time comes we will do the reverse, logic here
-        #       and calling this method from freenasUI.
-        return send_mail(**message)
+
+        syslog.openlog(logoption=syslog.LOG_PID, facility=syslog.LOG_MAIL)
+        interval = message.get('interval')
+        if interval is None:
+            interval = timedelta()
+        else:
+            interval = timedelta(seconds=interval)
+
+        sw_name = self.middleware.call_sync('system.info')['version'].split('-', 1)[0]
+
+        channel = message.get('channel')
+        if not channel:
+            channel = sw_name.lower()
+        if interval > timedelta():
+            channelfile = '/tmp/.msg.%s' % (channel)
+            last_update = datetime.now() - interval
+            try:
+                last_update = datetime.fromtimestamp(os.stat(channelfile).st_mtime)
+            except OSError:
+                pass
+            timediff = datetime.now() - last_update
+            if (timediff >= interval) or (timediff < timedelta()):
+                # Make sure mtime is modified
+                # We could use os.utime but this is simpler!
+                with open(channelfile, 'w') as f:
+                    f.write('!')
+            else:
+                raise CallError('This message was already sent in the given interval')
+
+        if not config:
+            config = self.middleware.call_sync('mail.config')
+        to = message.get('to')
+        if not to:
+            to = [
+                self.middleware.call_sync(
+                    'user.query', [('username', '=', 'root')], {'get': True}
+                )['email']
+            ]
+            if not to[0]:
+                raise CallError('Email address for root is not configured')
+
+        def read_json():
```

Comment:
We have plans to avoid reading entire attachments into memory, right?

Reply:
Thats why I limited it to 50MB.